hack/tools/prowjob-gen/generator.go

```diff
@@ -0,0 +1,164 @@
+		createdFiles: map[string]bool{},
+	}
+
+	var err error
+	g.templates, err = template.New("").
+		Funcs(g.templateFunctions()).
+		ParseGlob(templatesDir + "/*.yaml.tpl")
+	if err != nil {
+		return nil, err
+	}
+
+	return g, err
+}
+
+type generator struct {
+	templates    *template.Template
+	config       Config
+	outputDir    string
+	createdFiles map[string]bool
+}
+
+// generate executes every template for every branch and writes the result to a
+// file in outputDir.
+func (g *generator) generate() error {
+	for _, tpl := range g.config.Templates {
+		for branch := range g.config.Branches {
+			out, err := g.executeTemplate(branch, tpl.Name)
+			if err != nil && !errors.Is(err, os.ErrNotExist) {
+				return errors.Wrapf(err, "Generating prowjobs for template %s", tpl.Name)
+			}
+
+			fileName := fmt.Sprintf(tpl.Format, strings.ReplaceAll(branch, ".", "-"))
+			filePath := filepath.Clean(path.Join(g.outputDir, fileName))
+			if err := os.WriteFile(filePath, out.Bytes(), 0644); err != nil { //nolint:gosec
+				return errors.Wrapf(err, "Writing prowjob to %q", filePath)
+			}
+
+			g.createdFiles[fileName] = true
+		}
+	}
+	return nil
+}
+
+// cleanup deletes files which have the generatedFileHeader and had not been updated
+// during generate.
+func (g *generator) cleanup() error {
+	entries, err := os.ReadDir(g.outputDir)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		if _, ok := g.createdFiles[entry.Name()]; ok {
+			continue
+		}
+
+		if entry.IsDir() {
+			continue
+		}
+
+		path := filepath.Clean(path.Join(g.outputDir, entry.Name()))
+		data, err := os.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		if strings.HasPrefix(string(data), generatedFileHeader) {
+			klog.Infof("Deleting file %s", entry.Name())
+			if err := os.Remove(path); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// executeTemplate executes a previously parsed template with the data for a specific branch.
+func (g *generator) executeTemplate(branch, templateName string) (*bytes.Buffer, error) {
+	klog.Infof("executing template %q for branch %q", templateName, branch)
+
+	data := map[string]interface{}{
+		"branch":   branch,
+		"config":   g.config.Branches[branch],
+		"versions": g.config.VersionsMapper,
+	}
+
+	var out bytes.Buffer
+
+	// Write yaml comment as header to indicate this file got generated.
+	out.WriteString(generatedFileHeader)
+
+	if err := g.templates.ExecuteTemplate(&out, templateName, data); err != nil {
+		return nil, errors.Wrapf(err, "Executing template %q for branch %q", templateName, branch)
+	}
+
+	return &out, nil
+}
+
+// templateFunctions returns the functions available inside of templates.
+func (g *generator) templateFunctions() template.FuncMap {
+	funcs := map[string]any{}
+	funcs["TrimPrefix"] = strings.TrimPrefix
+	funcs["TrimSuffix"] = strings.TrimSuffix
+	funcs["ReplaceAll"] = strings.ReplaceAll
+	funcs["last"] = last
+	return funcs
+}
+
+func last(list any) any {
```

Comment:
Can this be done with generics instead of reflection?

Reply:
I did not find a way to use generics here. May be because the template.FuncMap is a `map[string]any`

Reply:
Alright, no problem